### ISSUE TITLE
[Metal Direct] assert program buffer allocation > `dram_unreserved_end`

### ISSUE THREAD

**Issue by nsmithtt** [tenstorrent/tt-mlir#539]
Here are some additional details to start:

Two stage change:

- On the tt-metal side add a method to Program class to return the `kernels_buffer` member (`Buffer` object which has public `address() `function for getting address). Today, we're only interested in the address, but in the future we could add method for serializing programs.

- On our (tt-mlir runtime) side we really just want to assert that the program device buffer address (from change in previous bullet) doesn't collide with our tensors / carve out region for the tensors. 

Some additional context for this ticket:

`dram_unreserved_end` was recently added to system-desc as the top of our usable space, and above that is (conservative) reserved space we calculate as this variable.  So this task is to add an assert that confirms actual program kernels space (from metal's kernels_buffer Buffer address) is indeed above that dram_unreserved_end we estimated (otherwise we may accidentally put tensors in space metal uses for kernels, causing bad things to happen).  The idea works because this space in metal is allocated "top down" in DRAM (starting from the end) rather than bottom-up.

The simple assert will need to be in runtime after Program object is fully formed. May need to pull in systemdesc to places it hasn't been used before.

**Comment by kmabeeTT** [comment]
> I believe there are two separate instances of class Program. I am assuming I am working on /tt-mlir/build/runtime/tools/python/ttrt/runtime/tt_metal/impl/program/program.hpp
> 
> The other one is /tt-mlir/third_party/tt-metal/src/tt-metal/tt_metal/impl/program/program.hpp

@ddilbazTT Not quite, the first one is under "build" so it's not source code, but it looks like exact copy of the 2nd file which is the real source code (inside tt-metal project) that we need to modify. Since it's in tt-metal it requires PR and approval in their project, then some time before it can make it's way to tt-mlir project (we consume/use tt-metal at a specific commit, updated every week or so).  For now you can make/test all changes, get the PR approved/merged in metal, and then merge our side (tt-mlir runtime) after the tt-metal version we consume gets uplifted to include the needed changes.

**Comment by ddilbazTT** [comment]
I don't think kernels_buffer is set anywhere in the code right now. Where should it be set/ edited?

**Comment by kmabeeTT** [comment]
> I don't think kernels_buffer is set anywhere in the code right now. Where should it be set/ edited?

It's populated in `program.cpp` from tt-metal here, you can add a print statement to see the relative ordering:

```
this->kernels_buffer = std::make_shared<Buffer>...
```

This called from `Program::compile()` path... which at the moment I'm not sure what triggers (looking)... we hoped it was before this call in our `command_queue.cpp`, but maybe not:

```
::tt::tt_metal::EnqueueProgram(*cq, program, blocking);
```

**Comment by ddilbazTT** [comment]
Can confirm compile is never hit.

**Comment by ddilbazTT** [comment]
Summary of progress to explain problems with adding assert (second point in the issue).
On metal side, included the following in

**Comment by skhatriTT** [comment]
@nsmithtt What is our status on this one?

**Comment by nsmithtt** [comment]
Actually I have a PR to fix this soon.  Hopefully be eow
